render/render.c

```diff
@@ -194,6 +197,22 @@ void wlr_render_subtexture(struct wlr_render *rend, struct wlr_tex *tex,
 	glEnableVertexAttribArray(texcoord_loc);
 
 	glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+}
+
+void wlr_render_subtexture(struct wlr_render *rend, struct wlr_tex *tex,
+		int32_t tex_x1, int32_t tex_y1, int32_t tex_x2, int32_t tex_y2,
+		int32_t pos_x1, int32_t pos_y1, int32_t pos_x2, int32_t pos_y2, int32_t pos_z) {
+	assert(eglGetCurrentContext() == rend->egl->context);
+	DEBUG_PUSH;
+
+	float verts[] = {
+		pos_x1, pos_y1, pos_z,
+		pos_x2, pos_y1, pos_z,
+		pos_x1, pos_y2, pos_z,
+		pos_x2, pos_y2, pos_z,
```

Comment:
What's pos_z for? Both here and elsewhere.

Reply:
We could change the vertex layout to be two dimensional if this isn't important.

Reply:
I had an idea of using a z dimension to handle laying 2D surfaces on top of each other (so you don't have to draw them in order), but nothing actually uses it yet, and I'm not even 100% sure if it'll work like I expect it to.

Reply:
Hmm, interesting idea, but I'm not sure how well it'd work over just rendering the views in order. It wouldn't work with semitransparent views, right? Those would have to be rendered in a separate pass anyway afaik.

Reply:
Semi-transparent surfaces were actually what I had in mind.
Draw the transparent surface at z == 2 **then** draw an opaque surface at z == 1, and then I think the GPU will just handle blending them together properly.
This type of "optimisation" basically kills the chances for this to be implemented with a software renderer, though.

Reply:
I don't think gles2 actually has orthographic projection, so I don't think this trick will work, so I removed it.

Reply:
Projection is done by the fragment shader, isn't it? gles2 doesn't have any functions for generating or manipulating matricies, it just passes them along to to the shader.